### PR TITLE
[ADAM-1083] Cleaning up `org.bdgenomics.adam.models`.

### LINKED ISSUE #1083: Clean up packages
Documentation and packaging cleanup for:

- [ ] `org.bdgenomics.adam.models`
- [x] `org.bdgenomics.adam.projections`
- [ ] `org.bdgenomics.adam.rich` see #1263
- [x] `org.bdgenomics.adam.serialization`
- [x] `org.bdgenomics.adam.util`
- [x] `org.bdgenomics.adam.rdd` classes not cleaned up in #1051

### PR DESCRIPTION
**Pull request by fnothaft**
Along with #1263 and #1264, this resolves #1083.

* Removing unused org.bdgenomics.adam.models.ReadBucket class.
* Move org.bdgenomics.adam.models.ReferencePositionPair and org.bdgenomics.adam.models.SingleReadBucket in to org.bdgenomics.adam.rdd.read and make package private.
* Clean up duplicated methods and methods that were incorrectly in companion singleton for SequenceDictionary and ReadGroupDictionary.
* Removed all SamReader references.
* Make writable file headers private to ADAM.
* Eliminated manual VCF parsing code in SnpTable.
* Cleaned up scaladoc for all classes and singleton objects.
* Moved `NonoverlappingRegions` test code out of `InnerBroadcastRegionJoinSuite`.

### CODE REVIEW

File: adam-core/src/main/scala/org/bdgenomics/adam/models/NonoverlappingRegions.scala [hunk 1]
```diff
@@ -186,23 +207,56 @@ object NonoverlappingRegions {
...
     regionMap.get(regionable._1.referenceName).fold(Iterable[ReferenceRegion]())(_.regionsFor(regionable))
 
+  /**
+   * A quick filter, to find out if we even need to examine a particular input value for keying by
+   * nonoverlapping-regions.  Basically, reject the input value if its corresponding region is
+   * completely outside the hull of all the input-set regions.
+   *
+   * @param regionable The input value
+   * @tparam U
+   * @return a boolean -- the input value should only participate in the regionJoin if the return value
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/NonoverlappingRegions.scala]: `@return True if ...` might read better

File: adam-core/src/main/scala/org/bdgenomics/adam/models/Attribute.scala [hunk 2]
```diff
@@ -18,17 +18,20 @@
...
- * embedding the corresponding information in the type of 'value', because otherwise it'd be difficult
- * to extract the correct type for Byte and NumericSequence values.
+ * A wrapper around the attrTuple (key) and value pair seen in many formats.
  *
- * Roughly analogous to Picards SAMTagAndValue.
+ * Includes the attrTuple-type explicitly, rather than embedding the
+ * corresponding information in the type of 'value', because otherwise it'd be
+ * difficult to extract the correct type for Byte and NumericSequence values.
+ *
+ * This class is roughly analogous to Picard's SAMTagAndValue.
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/Attribute.scala]: Picard &rarr; htsjdk

File: adam-core/src/main/scala/org/bdgenomics/adam/models/Alphabet.scala [hunk 3]
```diff
@@ -59,30 +64,38 @@ trait Alphabet {
...
+   *
+   * @see reverseComplementExact
    */
   def reverseComplement(s: String, notFound: (Char => Symbol) = ((c: Char) => Symbol(c, c))) = {
     s.map(x => Try(apply(x)).getOrElse(notFound(x)).complement).reverse
   }
 
-  /** number of symbols in the alphabet */
+  /**
+   * The number of symbols in the alphabet
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/Alphabet.scala]: end with period

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala [hunk 4]
```diff
@@ -328,28 +377,55 @@ case class ReferenceRegion(
...
+   * @return True if the region is wholly contained within our region.
+   */
   def contains(other: ReferenceRegion): Boolean = {
     orientation == other.orientation &&
       referenceName == other.referenceName &&
       start <= other.start && end >= other.end
   }
 
+  /**
+   * Checks is our region overlaps (wholly or partially) another region.
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: is &rarr; if

**Comment by fnothaft** [comment]
Thanks for the review @heuermh! I've addressed your comments, rebased, and pushed.

**Comment by heuermh** [comment]
Thank you, @fnothaft!

**Pull request merged**